gk/main.c

```diff
@@ -16,22 +16,404 @@
+	/* IP flow information. */
+	struct ip_flow flow;
+
+	/* The state of the entry. */
+	unsigned state : 2;
+
+	union {
+		struct {
+			/* The time the last packet of the entry was seen. */
+			uint64_t last_packet_seen_at;
+			/* 
+			 * The priority associated to
+			 * the last packet of the entry.
+			 */
+			uint8_t last_priority;
+			/* 
+			 * The number of packets that the entry is allowed
+			 * to send with @last_priority without waiting
+			 * the amount of time necessary to be granted
+			 * @last_priority.
+			 */
+			uint8_t allowance;
+			/* 
+			 * The ID of the Grantor server to which packets to
+			 * @dst must be sent.
+			 */
+			int grantor_id;
+		} request;
+
+		struct {
+			/* When the granted capability expires. */
+			uint64_t cap_expire_at;
+			/* When @budget_byte is reset. */
+			uint64_t budget_renew_at;
+			/* 
+			 * When @budget_byte is reset, reset it to
+			 * @tx_rate_kb_cycle * 1024 bytes.
+			 */
+			int tx_rate_kb_cycle;
+			/* How many bytes @src can still send in current cycle. */
+			int budget_byte;
+			/*
+			 * The ID of the Grantor server to which packets to
+			 * @dst must be sent.
+			 */
+			int grantor_id;
+			/* When GK should send the next renewal to @grantor_id. */
+			uint64_t send_next_renewal_at;
+			/*
+			 * How many cycles (unit) GK must wait before
+			 * sending the next capability renewal request.
+			 */
+			uint64_t renewal_step_cycle;
+		} granted;
+
+		struct {
+			/*
+			 * When the punishment (i.e. the declined capability)
+			 * expires.
+			 */
+			uint64_t expire_at;
+		} declined;
+	} u;
+};
+
+static uint8_t
+integer_log_base_2(uint64_t delta_time)
+{
+	return (8 * sizeof(uint64_t) - __builtin_clzll((delta_time)) - 1);
+}
+
+/* 
+ * It converts the difference of time between the current packet and 
+ * the last seen packet into a given priority. 
+ */
+static uint8_t 
+priority_from_delta_time(uint64_t present, uint64_t past)
+{
+	uint64_t delta_time;
+
+	if (unlikely(present < past)) {
+		/*
+		 * This should never happen, but we handle it gracefully here 
+		 * in order to keep going.
+		 */
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: the present time smaller than the past time!\n");
+
+		return 0;
+	}
+
+	delta_time = (present - past) * picosec_per_cycle;
+	if (unlikely(delta_time < 1))
+		return 0;
+	
+	return integer_log_base_2(delta_time);
+}
+
+static inline void
+extract_packet_info(struct rte_mbuf *pkt, struct ipacket *packet)
+{
+	uint16_t ether_type;
+	struct ether_hdr *eth_hdr;
+	struct ipv4_hdr  *ip4_hdr;
+	struct ipv6_hdr  *ip6_hdr;
+
+	eth_hdr = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	ether_type = rte_be_to_cpu_16(eth_hdr->ether_type);
+
+	switch (ether_type) {
+	case ETHER_TYPE_IPv4:
+
+		ip4_hdr = rte_pktmbuf_mtod_offset(pkt, 
+					struct ipv4_hdr *, 
+					sizeof(struct ether_hdr));
+
+		packet->flow.proto = ETHER_TYPE_IPv4;
+		packet->flow.f.v4.src = rte_be_to_cpu_32(ip4_hdr->src_addr);
+		packet->flow.f.v4.dst = rte_be_to_cpu_32(ip4_hdr->dst_addr);
```

Comment:
I used to converting all the network order to cpu order if we need to store them in memory, where it uses cpu order. The case for the IPv6 keeps the network order, since I did find a way to do the conversion. If we decide to use the network order, then I can keep this guideline consistent, i.e., storing packet/flow information in CPU also as network order.